src/search_decoder.py

```diff
@@ -94,91 +96,96 @@ def eval(self, alpha=1.0):
+
+        # starting node -  hidden vector, previous node, word id, logp, length
+        node = self.BeamSearchNode(decoder_hidden, None, decoder_input, 0, 1)
+        nodes = PriorityQueue()
+
+        # start the queue
+        nodes.put((-node.eval(), node))
+        qsize = 1
+
+        # start beam search
+        while True:
+            # give up when decoding takes too long
+            if qsize > 2000: break
+
+            # fetch the best node
+            score, n = nodes.get()
+            decoder_input = n.wordid
+            decoder_hidden = n.h
+
+            if n.wordid.item() == eos and n.prevNode != None:
+                endnodes.append((score, n))
+                # if we reached maximum # of sentences required
+                if len(endnodes) >= number_required:
+                    break
+                else:
+                    continue
+
+            # Forward pass through decoder
+            # Transform speaker_id from int into tensor with shape=(1, 1)
+            speaker_input = torch.LongTensor([[speaker_id]])
+            speaker_input = speaker_input.to(device)
+            # decode for one step using decoder
+            decoder_output, decoder_hidden = self.decoder(decoder_input, speaker_input, decoder_hidden, encoder_outputs)
+            # PUT HERE REAL BEAM SEARCH OF TOP
+            log_prob, indexes = torch.topk(decoder_output, self.beam_width)
+            nextnodes = []
+
+            for new_k in range(self.beam_width):
+                decoded_t = indexes[0][new_k].view(1, -1)
+                log_p = log_prob[0][new_k].item()
+
+                node = self.BeamSearchNode(decoder_hidden, n, decoded_t, n.logp + log_p, n.leng + 1)
+                score = -node.eval()
+                nextnodes.append((score, node))
+
+            # put them into queue
+            for i in range(len(nextnodes)):
+                score, nn = nextnodes[i]
+                nodes.put((score, nn))
+                # increase qsize
+            qsize += len(nextnodes) - 1
+
+        # choose nbest paths, back trace them
+        if len(endnodes) == 0:
+            endnodes = [nodes.get() for _ in range(topk)]
+
+        # Record token and score
+        all_tokens = []
+        all_scores = []
+        for score, n in sorted(endnodes, key=operator.itemgetter(0)):
+            # back trace
+            tokens = []
+            scores = []
+            while n.prevNode != None:
+                n = n.prevNode
+                tokens.append(n.wordid)
+                scores.append(n.logp)
+            all_tokens.append(tokens[::-1][1:])
+            all_scores.append(scores[::-1][1:])
+        idx = self.random_pick(len(all_tokens))
+        return all_tokens[idx], all_scores[idx]
+
+    def random_pick(self, num):
+        weight = [i for i in range(num,0,-1)]
+        num_exp = np.exp(weight)
+        prob = num_exp / num_exp.sum(0)
+        return np.random.choice(num, 1, p=prob)[0]
```

Comment:
The way of picking an output is pretty naive, e.g. there 3 choices, the weight will be the square of its index, [9, 4, 1]. We can improve this later if we want.